### PR TITLE
Add reserved namespace bindings.

### PR DESCRIPTION
**Pull request by wt**
This adds xml and xmlns namespace bindings. These are defined at https://www.w3.org/TR/xml-names11/#xmlReserved.

This is a a start of the work needed for #464

### CODE REVIEW

File: src/name.rs
Comment (Mingun) [inline src/name.rs:543]: I would like to see documentation on one of this consts about special processing of those prefixes with the link to the specification (https://www.w3.org/TR/xml-names11/#xmlReserved). Maybe just an adopted copy fro that specification.

Also, can we change type of these consts to `Namespace<'static>`?

I also prefer to have consts declared at the start of `impl` block, or you can declare them outside of `impl` block.

File: src/name.rs
Comment (Mingun) [inline src/name.rs:552]: This looks better:
```suggestion
            Some(Prefix(b"xml")) => ResolveResult::Bound(Self::XML_NAMESPACE),
            Some(Prefix(b"xmlns")) => ResolveResult::Bound(Self::XMLNS_NAMESPACE),
            Some(p) => ResolveResult::Unknown(p.into_inner().to_vec()),
```
Also, need to add a TODO comment, that this will not work with UTF-16 and ISO_2022_JP encodings, because namespaces stored in document encoding. As written it will work only because all other [supported encodings](https://docs.rs/encoding_rs/0.8.31/encoding_rs/index.html#no-extensibility-by-design) are ASCII-compatible.

File: src/name.rs
Comment (Mingun) [inline src/name.rs:817]: I would prefer to create a submodule `builtin_prefix` with `xml` and `xmlns` tests, and additional tests for overriding (maybe with TODOs that overriding rules should be changed)

File: src/name.rs
Comment (Mingun) [inline src/name.rs:825]:
```suggestion
            resolver.resolve(name, b"", true),
```
Here and everywhere.

**Review by Mingun — CHANGES_REQUESTED** [review]
Looks good, some minor notes for maintainability. Also, please add an entry to `Changelog.md` and probably one test to unit tests with a whole non-UTF-8 XML to ensure, that our expectations really works

**Comment by wt** [comment]
Okay, so I have declared the predefined namespaces in what I believe is a better way and added some of the docs you asked for. Is this heading in the right direction? By my calculation, the following still needs to be done:

- [ ] tests for overriding xml namespace (were there other override tests you were looking for?)
- [ ] ???

File: src/name.rs
Comment (wt) [reply src/name.rs]: I have pulled the definition of the prefixes and the actual mapping to be next to one another in the code. I have used the Prefixes in a more abstract manner that I think obviates the needs to TODO this code since changing the implementation of Prefix is what appears to be needed to fix the non-ascii compat encodings. Please let me know if you agree.

File: src/name.rs
Comment (wt) [reply src/name.rs]: I documented the reserved namespaces and changed their types. I also moved them to lazily initalized static outside of the impl. I did this as I wanted to have a HashMap that mapped the prefix to the Namespace. I also wanted to define the prefix and appropriate namespace in the same part of the code. I ended up using the once_cell, which is due to be incorporated in core at some point (https://doc.rust-lang.org/core/cell/struct.OnceCell.html#method.eq). The core support for OnceCell is currently experimental, so I hope this (using the crate) is acceptable until it's stabilized as I think it's a much better organization for the code.

Comment (wt) [reply src/name.rs]: https://doc.rust-lang.org/core/cell/struct.LazyCell.html is the location of the Lazy related stuff in core.

**Review by Mingun — CHANGES_REQUESTED** [review]
For override tests: check the behavior of other mature XML libraries, for example, [Java](https://docs.oracle.com/en/java/javase/19/docs/api/java.xml/module-summary.html). Then write the tests in `tests/namespaces.rs`. For example, check this XML document:
```xml
<root xml:attribute="ok">
  <xml:element/>

  <override xmlns:xml="overrided non-overradable prefix" xml:attribute="oh!">
    <xml:element/>
  </override>

  <!-- remove standard prefix? -->
  <reset xmlns:xml="" xml:attribute="eh?">
    <xml:element/>
  </reset>
</root>
```
How other libraries will parse that document? If it will be parsed, what the `QName`s will be returned?

It also would be excellent if you add test to `tests/encoding.rs` with some non-UTF-8 document and check that we correctly process `xml:` and `xmlns:` attributes:
- check that we detect and use non-UTF-8 encoding by querying `.decoder()`
- check that we element and attribute with undeclared prefix resolved correctly (well, we can even made `XML_NAMESPACE` and `XMLNS_NAMESPACE` our public API -- as a `Namespace` constants)
- at the same time check, that other undeclared prefix will resolved as unknown

**Comment by wt** [comment]
Just so we know, the std versions of the oncecell stuff that I used in this PR is being stabilized as we speak. Part of the work landed in 1.70. The lazy initialized types have not yet landed in stable. However, that may happen in one of the next couple releases. My plan is to update this PR when that lands.

**Comment by dralley** [comment]
@wt Even once it's merged, I don't know that we want to immediately bump the minimum supported Rust version up to the latest one.  If that is the only thing holding this PR up, if it's ready and @Mingun likes it, then I would say just merge it sooner rather than later, and we can drop the dependency later on.

**Comment by wt** [comment]
Can we just use the once_cell crate via [this](https://docs.rs/once_cell/latest/once_cell/sync/struct.Lazy.html) then? This crate will provide backward compatibility until this is stabilized, and it has been around for while.

**Comment by dralley** [comment]
>Can we just use the once_cell crate via [this](https://docs.rs/once_cell/latest/once_cell/sync/struct.Lazy.html) then? This crate will provide backward compatibility until this is stabilized, and it has been around for while.

Yup that's exactly what I'm suggesting.  Use the `once_cell` crate, so that we don't have to wait for it to hit the stdlib and/or immediately bump the MSRV when it does.

File: src/name.rs
Comment (dralley) [reply src/name.rs:830]: @wt One other remaining review comment ^^

**Comment by dralley** [comment]
Also this should have a changelog entry

**Comment by dralley** [comment]
Looks like `once_cell` 1.18 has an MSRV of 1.60 whereas the current MSRV for this library is 1.52.  

I'm fine with bumping it up to 1.60  (released April 2022), that is still pretty reasonable.  Most linux distros will have something newer which is really the main constraint.  But that should get a changelog entry also.

**Comment by wt** [comment]
Okay, so I hammered out a test xml parser with Xerces. It does in fact map to the well known namespaces to their appropriate URIs the same way this change does.

This doc:
```
<document
    xmlns:xsi='http://www.w3.org/2001/XMLSchema-instance'
    xsi:noNamespaceSchemaLocation='document.xsd' 
    xml:id='fjdlksfjds'>
</document>
```

produces the following parsed attrs for the document tag:
```
--> [document], Child Element Count = 0
attr: http://www.w3.org/XML/1998/namespace:id
attr: http://www.w3.org/2000/xmlns/:xsi
attr: http://www.w3.org/2001/XMLSchema-instance:noNamespaceSchemaLocation
```

Also, the failing tests appear to not be caused by my change. Do you agree with this?

I am also trying to figure out how to address your comment about checking the encoding. Can you help me figure out what it is that you want there?

**Comment by wt** [comment]
FWIW, Xerces also does not allow overriding the xml namespace to another URI or to no URI. I guess this is the behavior we would want in quick-xml?

And I added a changelog.

Here's a possible task list to finish this change:

- [ ] disallow overriding the xml and xmlns namespace prefixes
- [ ] add integration test that tests a full document
- [ ] properly handle non-ascii compatible encodings

Does that cover everything?

**Review by Mingun — CHANGES_REQUESTED** [review]
> Can we just use the once_cell crate via [this](https://docs.rs/once_cell/latest/once_cell/sync/struct.Lazy.html) then? This crate will provide backward compatibility until this is stabilized, and it has been around for while.

There is no need to use hashmap and, as a result, `once_cell`, or `Lazy`, at all. Just two entries can be just hardcoded in place of use via `match`, or you can use `&[(&str, &str)]` constant and use linear search -- this will be faster that any map on such sizes and will not require lazy construction and additional dependencies.

And I still prefer to test each well-known prefix individually (one test for prefix, you can write a small macro if you don't like to duplicate code).

**Comment by Mingun** [comment]
Those failures appeared since serde v1.0.181 due to https://github.com/serde-rs/serde/pull/2505:
```
failures:
    enum_::adjacently_tagged::nested_struct::elements
    enum_::adjacently_tagged::newtype::elements
    enum_::adjacently_tagged::struct_::elements
    enum_::adjacently_tagged::tuple_struct::elements
    enum_::adjacently_tagged::unit::elements
```
I've create https://github.com/tafia/quick-xml/issues/630 for investigation.

**Comment by wt** [comment]
The MSRV test is weird. Why is is failing for 1.52 when the min stated version is 1.60 in this change?

**Comment by dralley** [comment]
>The MSRV test is weird. Why is is failing for 1.52 when the min stated version is 1.60 in this change?

It's hardcoded https://github.com/tafia/quick-xml/blob/master/.github/workflows/rust.yml#L17

But I suppose it's not necessary to change if you go w/ the linear search approach.  With only a small number of strings to check, @Mingun is correct that it's not really going to be worth the Hashmap

**Comment by wt** [comment]
I'll rewrite without the HashMap stuff and change the tests to be now explicit.

Do y'all agree that the semantics around using the xml and xmlns prefixes should be enforced as they are in Java and Xerces?

**Comment by wt** [comment]
I just pushed a new change that uses no HashMap. Please let me know if this addresses the concerns on the design.

Also, I think this makes my task list above accurate. Are we in agreement on that?

File: src/name.rs
Comment (dralley) [inline src/name.rs:415]: I'd just call it `RESERVED_NAMESPACES`

**Review by Mingun — COMMENTED** [review]
> FWIW, Xerces also does not allow overriding the xml namespace to another URI or to no URI. I guess this is the behavior we would want in quick-xml?

What action Xerces did when it encounters such situation? It returns error while parsing or just does nothing?

> - [ ] disallow overriding the xml and xmlns namespace prefixes

With the current implementation this is already done if our desired action is "do nothing". But if Xerces returns an error then I think we should add a setting to replicate Xerces behavior (a setting "do nothing / produce error").

> - [ ] add integration test that tests a full document

Well, if you do that, fine, but I'm also OK to proceed without it.

> - [ ] properly handle non-ascii compatible encodings

I think that this task will impossible to implement here, so do not bother about it

> I just pushed a new change that uses no HashMap. Please let me know if this addresses the concerns on the design.

Yes, that's much better. Although in my opinion, even the constant is superfluous, readability is better without it (i.e. if you embed everything in `match`):
```rust
    /// Constants in `match` define the [reserved namespaces] for the xml standard.
    ///
    /// The prefix `xml` is by definition bound to the namespace name
    /// `http://www.w3.org/XML/1998/namespace`. It may, but need not, be declared, and must not be
    /// undeclared or bound to any other namespace name. Other prefixes must not be bound to this
    /// namespace name, and it must not be declared as the default namespace.
    ///
    /// The prefix `xmlns` is used only to declare namespace bindings and is by definition bound
    /// to the namespace name `http://www.w3.org/2000/xmlns/`. It must not be declared or
    /// undeclared. Other prefixes must not be bound to this namespace name, and it must not be
    /// declared as the default namespace. Element names must not have the prefix `xmlns`.
    ///
    /// [reserved namespaces]: https://www.w3.org/TR/xml-names11/#xmlReserved
    #[inline]
    fn maybe_unknown(prefix: Option<Prefix>) -> ResolveResult<'static> {
        match prefix {
            Some(Prefix(b"xml")) => {
                ResolveResult::Bound(Namespace(b"http://www.w3.org/XML/1998/namespace"))
            }
            Some(Prefix(b"xmlns")) => {
                ResolveResult::Bound(Namespace(b"http://www.w3.org/2000/xmlns/"))
            }

            Some(p) => ResolveResult::Unknown(p.into_inner().to_vec()),
            None => ResolveResult::Unbound,
        }
    }
```

**Comment by Mingun** [comment]
Oh, and you can rebase on current master to not suffer from failing tests

**Comment by wt** [comment]
rebased

**Comment by dralley** [comment]
Do you happen to know how libxml2 and/or expat handles them?

**Comment by wt** [comment]
I don't know how the other libs work.

**Comment by wt** [comment]
Looks like expat does a similar validation around [here](https://github.com/libexpat/libexpat/blob/20293f2ad29d5a5ebe126174e5bb961962540b81/expat/lib/xmlparse.c#L3834). By my reading errors are returned in the same cases as for Xerces.

File: src/name.rs
Comment (wt) [reply src/name.rs]: I'm planning on moving the logic depending on our agreement on the error behavior. I'm thinking this change may not use this constant, so I am going to try to get agreement on the error behavior and then rewrite with that in mind.

**Comment by wt** [comment]
Alright, this newest iteration completely changes how the implementation to insert the xml and xmlns namespace entries in the root namespace resolver. I have not yet written tests to cover all the new semantics when pushing a new namespace context, so this change is in progress.

Having said that, can y'all please make sure this implementation is heading in an acceptable direction?

**Comment by wt** [comment]
rebased

**Comment by dralley** [comment]
@wt There's are a few small compilation errors to address.

@Mingun Does this implementation strategy look reasonable to you?

**Comment by wt** [comment]
Fixed the lint issues for the compilation. I ran the individual tests, but no `cargo test` after the previous rebase.

File: src/errors.rs
Comment (dralley) [inline src/errors.rs:15]: Is there a particular reason to make separate structs instead of directly embedding the strings in the enum variants?  I don't think we use this pattern in many, if any, places.

File: src/name.rs
Comment (dralley) [inline src/name.rs:456]: This is a breaking change (which is fine) but it should probably have its own callout in the changelog.

File: src/name.rs
Comment (dralley) [inline src/name.rs:428]: I have *not* done a thorough search and @Mingun is the expert on this code not me, but it seems that the buffer is always? used alongside the `NamespaceResolver`, in which case, is there a reason it can't be part of the `NamespaceResolver` rather than being passed around?

(I may have missed somewhere where it was being used elsewhere)

Comment (wt) [reply src/name.rs:428]: I was wondering the same, but I really didn't have an interest in making a change to that design as part of this change.

**Comment by wt** [comment]
Okay, I think the latest version is ready. I've added tests for adding namespace contexts to the resolver. I added these specific tests:

* xml prefix explicit sets work when the uri is correct
* xml prefix sets fail when setting to incorrect url
* setting other prefixes to the xml uri fails
* setting xmlns prefix explicitly fails
* setting other prefixes to the xmlns uri fails
* setting xml to the xmlns prefix fails (I added this as the logic is overlapping in a way that is easy to biff)

I believe that this is ready to a review. @Mingun would you mind taking a look? I would love to get this landed soon.

File: src/errors.rs
Comment (wt) [reply src/errors.rs]: I've collapsed them into a single struct. I think improving the errors should probably be left for another change. I hope this addresses your concern.

File: src/name.rs
Comment (wt) [reply src/name.rs:456]: Does this require a major version bump? Also, should I add a completely separate entry for the return type change, or should I add it to the existing entry?

Comment (Mingun) [reply src/name.rs:456]: No, `NamespaceResolver` is a private struct, so this is not a breaking change

Comment (wt) [reply src/name.rs:456]: Given that, I'm assuming that this doesn't need to be cashed out in the changelog.

Comment (dralley) [reply src/name.rs:456]: yes, apologies.  oversight on my part

**Comment by Mingun** [comment]
Don't worry, I don't forgot about this. I was on vacation so tracked issues very lazily. I'll look during this week.

**Comment by Mingun** [comment]
Thanks for working on this PR and investigating all stuff about handling this in other libraries!

I'm not finishing my review yet. Instead of giving feedback in the form of comments, I decided that it would be easier to implement my vision myself. Changes I have broken down into several commits so you can see the reason for them.

I structurized tests and added missing parts (what became apparent after structuring). I also simplified checks and now, I think, it is more easily to understand the code.

I plan to finish review next week.

**Comment by wt** [comment]
I'm glad you find some improvement. I'm not picky as long as the functionality is there. When do you think this will land in the main branch?

**Comment by Mingun** [comment]
I think this week

File: src/name.rs
Comment (Mingun) [reply src/name.rs:431]: > I have _not_ done a thorough search and @Mingun is the expert on this code not me, but it seems that the buffer is always? used alongside the `NamespaceResolver`, in which case, is there a reason it can't be part of the `NamespaceResolver` rather than being passed around?

Yes, I also noticed this and will change this in that PR

**Review by Mingun — APPROVED** [review]
Well, I think, it is done. I'll merge it next day if there will be no objections

**Pull request merged**
